internal/keyrotation/rotate.go

```diff
@@ -62,29 +62,39 @@ func (s *Server) handleRotateKeys(ctx context.Context) http.HandlerFunc {
 func (s *Server) doRotate(ctx context.Context) error {
 	metrics := s.env.MetricsExporter(ctx)
 
-	_, allowed, err := s.revisionDB.GetAllowedRevisionKeys(ctx)
+	effectiveID, allowed, err := s.revisionDB.GetAllowedRevisionKeys(ctx)
 	if err != nil {
 		return fmt.Errorf("rotate-keys unable to read revision keys: %w", err)
 	}
 
 	// First allowed is newest due to sql orderby.
+	var previousCreated time.Time
 	if len(allowed) == 0 || time.Since(allowed[0].CreatedAt) >= s.config.NewKeyPeriod {
-		if _, err := s.revisionDB.CreateRevisionKey(ctx); err != nil {
+		key, err := s.revisionDB.CreateRevisionKey(ctx)
+		if err != nil {
 			return fmt.Errorf("failed to create revision key: %w", err)
 		}
+		effectiveID = key.KeyID
+		previousCreated = key.CreatedAt
 		metrics.WriteInt("revision-keys-created", true, 1)
+	} else {
+		previousCreated = allowed[0].CreatedAt
 	}
 
 	var result error
 	deleted := 0
 	for _, key := range allowed {
-		if time.Since(key.CreatedAt) < s.config.DeleteOldKeyPeriod {
+		if key.KeyID == effectiveID {
 			continue
 		}
+		if time.Since(previousCreated) < s.config.DeleteOldKeyPeriod {
+			continue // A key is not safe to delete until the newer one was effective for the period.
```

Comment:
I think in this path we still want to update previousCreated

this key isn't ready to be deleted, but it can still push the timestamp down for the next in line.